### ISSUE TITLE
[no-throw-literal] False positive with nullish coalescing in throw statement

### ISSUE THREAD

**Issue by jsone-studios** [typescript-eslint/typescript-eslint#2642]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-throw-literal": "error"
  }
}
```

```TS
let error: Error | undefined
throw error ?? new Error()
```

```JSON
{
  "compilerOptions": {
    "target": "es2018",
    "module": "commonjs",
    "strict": true,
    "moduleResolution": "node",
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "allowJs": true,
    "baseUrl": ".",
    "paths": {
      "@/*": ["src/*"]
    },
    "lib": ["es2018"],
    "outDir": "dist"
  },
  "include": [
    "src/**/*.ts",
  ],
  "exclude": ["node_modules"]
}
```

**Expected Result**

No error is reported, as in this case the nullish coalescing operator is always returning something of type `Error`

**Actual Result**

An error is reported by the `@typescript-eslint/no-throw-literal` rule: `Expected an error object to be thrown`

**Additional Info**

It seems the nullish coalescing operator needs special handling in this rule, similar to how intersection and union types are handled (see #2085). I try to provide a PR the next days. 

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.4.0` |
| `@typescript-eslint/parser`        | `4.4.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.10.0` |
| `node`                             | `12.16.2` |

**Comment by bradzacher** [comment]
Should be pretty easy to fix - happy to accept a PR!

The bug is specifically here.
https://github.com/typescript-eslint/typescript-eslint/blob/6bc93257ec876214743a165093b6666d713379f6/packages/eslint-plugin/src/rules/no-throw-literal.ts#L83-L86

The logic is kind of off for LogicalExpressions. Right now it introspects the LogicalExpression and favours the left side over the right, and inspects just one of them.
I think it probably should either consider the logical as a whole, or it should check all parts of the logical.

Right now it will work incorrectly for the following cases:
```
declare const nullishError: Error | undefined;

throw nullishError ?? new Error();               // false-positive
throw nullishError || new Error();               // false-positive
throw 'literal' || new Error();                  // false-negative
throw nullishError ? nullishError : new Error(); // false-positive
```
